test/js_test.go

```diff
@@ -2243,3 +2247,509 @@ func TestJetStream_UnsubscribeDeleteNoPermissions(t *testing.T) {
+	n := 3
+	replicas := []int{1, 3}
+
+	t.Run("pull sub", func(t *testing.T) {
+		for _, r := range replicas {
+			t.Run(fmt.Sprintf("n=%d r=%d", n, r), func(t *testing.T) {
+				stream := &nats.StreamConfig{
+					Name:     fmt.Sprintf("foo-r%d", r),
+					Replicas: r,
+				}
+				withJSClusterAndStream(t, fmt.Sprintf("PULLR%d", r), n, stream, testJetStream_ClusterReconnectPullSubscriber)
+			})
+		}
+	})
+
+	t.Run("qsub durable", func(t *testing.T) {
+		for _, r := range replicas {
+			t.Run(fmt.Sprintf("n=%d r=%d", n, r), func(t *testing.T) {
+				stream := &nats.StreamConfig{
+					Name:     fmt.Sprintf("bar-r%d", r),
+					Replicas: r,
+				}
+				withJSClusterAndStream(t, fmt.Sprintf("QSUBR%d", r), n, stream, testJetStream_ClusterReconnectDurableQueueSubscriber)
+			})
+		}
+	})
+
+	t.Run("sub durable", func(t *testing.T) {
+		for _, r := range replicas {
+			t.Run(fmt.Sprintf("n=%d r=%d", n, r), func(t *testing.T) {
+				stream := &nats.StreamConfig{
+					Name:     fmt.Sprintf("quux-r%d", r),
+					Replicas: r,
+				}
+				withJSClusterAndStream(t, fmt.Sprintf("SUBR%d", r), n, stream, testJetStream_ClusterReconnectDurablePushSubscriber)
+			})
+		}
+	})
+}
+
+func testJetStream_ClusterReconnectPullSubscriber(t *testing.T, subject string, srvs ...*jsServer) {
+	var (
+		recvd         int
+		srvA          = srvs[0]
+		totalMsgs     = 20
+		durable       = nats.Durable("d1")
+		reconnected   = make(chan struct{}, 2)
+		reconnectDone bool
+	)
+	nc, err := nats.Connect(srvA.ClientURL(),
+		nats.ReconnectHandler(func(nc *nats.Conn) {
+			// t.Logf("Reconnected!")
+			reconnected <- struct{}{}
+
+			// Bring back the server after the reconnect event.
+			if !reconnectDone {
+				reconnectDone = true
+				srvA.Restart()
+			}
+		}),
+	)
+	if err != nil {
+		t.Error(err)
+	}
+	defer nc.Close()
+
+	js, err := nc.JetStream()
+	if err != nil {
+		t.Error(err)
+	}
+
+	for i := 0; i < 10; i++ {
+		payload := fmt.Sprintf("i:%d", i)
+		js.Publish(subject, []byte(payload))
+	}
+
+	sub, err := js.SubscribeSync(subject, durable, nats.Pull(1))
+	if err != nil {
+		t.Error(err)
+	}
+
+	for i := 10; i < totalMsgs; i++ {
+		payload := fmt.Sprintf("i:%d", i)
+		js.Publish(subject, []byte(payload))
+	}
+
+	ctx, done := context.WithTimeout(context.Background(), 10*time.Second)
+	defer done()
+
+Next:
+	for recvd != totalMsgs {
+		select {
+		case <-ctx.Done():
+			t.Fatalf("Timeout waiting for messages, expected: %d, got: %d", totalMsgs, recvd)
+		default:
+		}
+
+		pending, _, _ := sub.Pending()
+		if pending == 0 {
+			err = sub.Poll()
+			if err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+		}
+		msg, err := sub.NextMsg(2 * time.Second)
+		if err != nil {
+			// t.Logf("Error fetching next message on %v: %v", sub.Subject, err)
+			continue Next
+		}
+
+		// Server will shutdown after a couple of messages which will result
+		// in empty messages with an status unavailable error.
+		if len(msg.Data) == 0 && msg.Header.Get("Status") == "503" {
```

Comment:
When a server restarts in cluster mode, these empty messages are very common.  I think maybe we need a new error type here to handle temporary unavailable state instead of checking from the header.

Reply:
Would this be a bug when there is still a quorum available?